File: trunk/vs2010/Shared/PlateWork/ActivePlateFactory.cs
using System;
using System.Linq;

namespace BioNex.Shared.PlateWork
{
    public abstract class ActivePlateFactory
    {
        // ----------------------------------------------------------------------
        // properties.
        // ----------------------------------------------------------------------
        protected Worklist Worklist { get; set; }
        public int NumberOfPlatesToCreate { get; set; }
        public int PlateInstanceIndex { get; set; }

        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        public ActivePlateFactory( Worklist worklist)
        {
            Worklist = worklist;
            SetNumberOfPlatesToCreate();
            PlateInstanceIndex = 0;
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
        public abstract Type GetActivePlateType();

        public ActivePlate CreateActivePlate()
        {
            --NumberOfPlatesToCreate;
            // DKM 2011-10-05 I think whether or not using MutableString to solve the barcode-reassignment problem hinges
            //                on the creation of ActivePlate.  Worklist has a reference to Plates which have the MutableString.
            //                So the question is, when we create the ActivePlate object, are the Plate bits a copy?  Inside
            //                of the ActiveDestinationPlate constructor, it looks like the internal destination plate object
            //                is a new object.  Bummer.  Let's see what we can do in there.
            // DKM 2011-10-09 need to check if there are any more destination plates to create.  If not, bail, otherwise
            //                we'll go beyond the bounds of the dest plate array when calling CreateInstance.
            if (NumberOfPlatesToCreate < 0) {
                // DKM 2012-01-18 prevent possible OverflowException when CreatePlate is called over and over again
                NumberOfPlatesToCreate = 0;
                return null;
            }
            ActivePlate active_plate = ( ActivePlate)( Activator.CreateInstance( GetActivePlateType(), Worklist, PlateInstanceIndex));
            PlateInstanceIndex++;
            return active_plate;
        }

        public abstract void SetNumberOfPlatesToCreate();

        public abstract int NumberOfSimultaneousPlates();

        public virtual bool ReleaseActivePlate()
        {
            int active_plates_of_active_plate_type = ActivePlate.ActivePlates.Count( active_plate => active_plate.GetType() == GetActivePlateType());
            return active_plates_of_active_plate_type < NumberOfSimultaneousPlates();
        }

        public ActivePlate TryReleaseActivePlate()
        {
            return ReleaseActivePlate() ? CreateActivePlate() : null;
        }
    }

    public class ActiveSourcePlateFactory : ActivePlateFactory
    {
        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        public ActiveSourcePlateFactory( Worklist worklist)
            : base( worklist)
        {
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
        public override Type GetActivePlateType()
        {
            return typeof( ActiveSourcePlate);
        }

        public override void SetNumberOfPlatesToCreate()
        {
            NumberOfPlatesToCreate = Worklist.TransferOverview.Transfers.Select( t => t.SrcPlate).Distinct().Count();
        }

        public override int NumberOfSimultaneousPlates()
        {
            return 3;
        }

        public override bool ReleaseActivePlate()
        {
            return base.ReleaseActivePlate();
        }
    }

    public class ActiveDestinationPlateFactory : ActivePlateFactory
    {
        // ----------------------------------------------------------------------
        // members
        // ----------------------------------------------------------------------

        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        public ActiveDestinationPlateFactory( Worklist worklist)
            : base( worklist)
        {
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
        public override Type GetActivePlateType()
        {
            return typeof( ActiveDestinationPlate);
        }

        public override void SetNumberOfPlatesToCreate()
        {
            NumberOfPlatesToCreate = Worklist.TransferOverview.Transfers.Select( t => t.DstPlate).Distinct().Count();
        }

        public override int NumberOfSimultaneousPlates()
        {
            return 2;
        }

        public override bool ReleaseActivePlate()
        {
            return base.ReleaseActivePlate();
        }
    }
}

File: scheduler_framework/scheduler/active_plate_factory.py
"""
ActivePlateFactory - Factory pattern for creating ActivePlates.

Manages the creation and release of active plates based on
worklist requirements and concurrency limits.
"""

from abc import ABC, abstractmethod
from typing import Type, List
from .worklist import Worklist
from .active_plate import ActivePlate, ActiveSourcePlate, ActiveDestinationPlate


class ActivePlateFactory(ABC):
    """Abstract factory for creating active plates"""
    
    def __init__(self, worklist: Worklist):
        self.worklist = worklist
        self.number_of_plates_to_create = 0
        self.plate_instance_index = 0
        self._set_number_of_plates_to_create()
    
    @abstractmethod
    def get_active_plate_type(self) -> Type[ActivePlate]:
        """Get the type of active plate this factory creates"""
        pass
    
    @abstractmethod
    def _set_number_of_plates_to_create(self):
        """Set how many plates need to be created"""
        pass
    
    @abstractmethod
    def number_of_simultaneous_plates(self) -> int:
        """Get maximum number of plates that can be active simultaneously"""
        pass
    
    def create_active_plate(self) -> ActivePlate:
        """Create a new active plate"""
        if self.number_of_plates_to_create <= 0:
            return None
        
        self.number_of_plates_to_create -= 1
        plate_type = self.get_active_plate_type()
        active_plate = plate_type(self.worklist, self.plate_instance_index)
        self.plate_instance_index += 1
        return active_plate
    
    def release_active_plate(self) -> bool:
        """
        Check if a new active plate can be released.
        
        Returns True if a plate can be released based on concurrency limits.
        """
        active_plates_of_type = [
            ap for ap in ActivePlate._active_plates
            if isinstance(ap, self.get_active_plate_type())
        ]
        return len(active_plates_of_type) < self.number_of_simultaneous_plates()
    
    def try_release_active_plate(self) -> ActivePlate:
        """
        Try to release a new active plate.
        
        Returns the new active plate if one can be released, None otherwise.
        """
        if self.release_active_plate():
            return self.create_active_plate()
        return None


class ActiveSourcePlateFactory(ActivePlateFactory):
    """Factory for creating ActiveSourcePlate instances"""
    
    def get_active_plate_type(self) -> Type[ActivePlate]:
        return ActiveSourcePlate
    
    def _set_number_of_plates_to_create(self):
        """Set number based on unique source plates in transfers"""
        if self.worklist.transfer_overview:
            # Use barcode to identify unique plates since Plate objects aren't hashable
            unique_sources = set(
                transfer.source_plate.barcode 
                for transfer in self.worklist.transfer_overview.transfers
            )
            self.number_of_plates_to_create = len(unique_sources)
        else:
            self.number_of_plates_to_create = len(self.worklist.source_plates)
    
    def number_of_simultaneous_plates(self) -> int:
        """Maximum 3 source plates can be active simultaneously"""
        return 3


class ActiveDestinationPlateFactory(ActivePlateFactory):
    """Factory for creating ActiveDestinationPlate instances"""
    
    def get_active_plate_type(self) -> Type[ActivePlate]:
        return ActiveDestinationPlate
    
    def _set_number_of_plates_to_create(self):
        """Set number based on unique destination plates in transfers"""
        if self.worklist.transfer_overview:
            # Use barcode to identify unique plates since Plate objects aren't hashable
            unique_destinations = set(
                transfer.destination_plate.barcode 
                for transfer in self.worklist.transfer_overview.transfers
            )
            self.number_of_plates_to_create = len(unique_destinations)
        else:
            self.number_of_plates_to_create = len(self.worklist.destination_plates)
    
    def number_of_simultaneous_plates(self) -> int:
        """Maximum 2 destination plates can be active simultaneously"""
        return 2

